BREAKING: WebGPU support (MToonNodeMaterial)

0b5vr opened the pull request:
This PR supports WebGPURenderer by adding `MToonNodeMaterial`, which is a NodeMaterial port of `MToonMaterial`.

After the change is published, you can use three-vrm on WebGPU by using `MToonNodeMaterialLoaderPlugin`.
To use the loader, specify the loader in the option of `VRMLoaderPlugin`.
See the `webgpu-dnd.html` example for details.

### BREAKING

- `gltf.userData.vrmMToonMaterials` might have materials other than `MToonMaterial` (namely, `MToonNodeMaterial` in this case).
    - This should not affect many users since the interface of `MToonNodeMaterial` is almost identical to `MToonMaterial`.

### TODOs

- [x] Since `three-vrm-material-mtoon` imports `three/addons/nodes/Nodes.js`, the initial module loading time became longer when you are using certain CDNs like unpkg. We are currently seeking for a workaround.
    - See also: https://discourse.threejs.org/t/nodematerial-on-unpkg-slow-to-load-all-the-dependencies/63213
    - Replaced the CDN from unpkg to jsdelivr and the problem is resolved.
        - See: https://github.com/mrdoob/three.js/pull/28006
        - See: https://github.com/pixiv/three-vrm/pull/1385
- [x] Several type definitions on `@types/three` are patched by `patch-package`
    - We might want to fix by `three-ts-types` side before we ship this
- [x] The `Type of property 'cache' circularly references itself ...` error (TS2615) occurs
    - We might want to bump our TypeScript version to 5.1 or later
    - The issue is resolved, at least in the patch
        - See: https://github.com/three-types/three-ts-types/pull/744#issuecomment-2036440655

0b5vr committed:
deps: Bump three.js to r160

- The type parameter of `THREE.Object3D` is no longer required (used to be `THREE.Object3D<THREE.Event>`
- `THREE.UniformsUtils.merge` now has more appropriate typing which breaks the existing code, workarounded with `as any`

0b5vr committed:
wip feature: MToon for WebGPU

Built the MToon shader using NodeMaterial feature

I had to add `three/examples/jsm/nodes/Nodes.js` to rollup's `external` config. Is the solution legit?

We have a problem on the current `@types/three`, I had to patch-package. Some of them are already addressed as PR on `three-ts-types` repo

I applied temporary changes to several example html files, this should be reverted before merge

0b5vr committed:
examples: Change light intensity from 1 to PI

Starting from r155, the light unit is set to physically based one by default.
We need to multiply the directional light intensity by PI because of this change.

See: https://github.com/mrdoob/three.js/pull/26392

0b5vr committed:
refactor: add COMPAT comment on pre-r156 compats addressed in https://github.com/pixiv/three-vrm/pull/1304

also removed an impossible compat code path

0b5vr committed:
fix (WebGPU, MToon): Fix outline, it should scale in world space instead of local space

0b5vr committed:
fix (WebGPU, MToon): Fix transparentWithZWrite

`parameters` have `depthWrite = false` when it's transparent, we have to overwrite this

0b5vr committed:
fix: Fix outline width when model normal is not normalized

It seems model normal was multiplied by 100 in UniGLTF 1.27, 1.28. Low confidence

0b5vr committed:
fix: Ignore vertex color if it's defined

also emit a warning if the model contains vertex colors

0b5vr committed:
fix: (WebGPU, MToon): Fix parametricRim formula

0b5vr committed:
fix (WebGPU, MToon): Fix outline color calculation method

The outline was too bright when the outlineLightingMix is zero

0b5vr committed:
fix (WebGPU, MToon): set alpha to 1 when the material is transparent

This prevents opaque / cutout materials to go white

0b5vr committed:
feature (WebGPU, MToon): support screenCoordinate width outline and outlineWidthTexture

0b5vr committed:
fix (WebGPU, MToon): fix matcap and parametric rim intensity when rimLightingMix factor is 1

it should be multiplied by (1.0 / PI)

0b5vr committed:
fix (WebGPU, MToon): Fix outline color alpha

outline didn't respect alpha when outlineLightingMixFactor is not 1

0b5vr committed:
improve (WebGPU, MToon): improve customProgramCacheKey

outlineWidthMode is not required when it's not outline pass

also add comma

0b5vr committed:
improve (WebGPU, MToon): remove outlineWidthMode from customProgramCacheKey

it's included in cacheKey in either way.
See: https://github.com/mrdoob/three.js/blob/r160/examples/jsm/renderers/common/RenderObject.js#L115

0b5vr committed:
improve (WebGPU, MToon): Improve code around parametric rim

We should have the parametric rim result to a temporary variable

0b5vr committed:
improve: Improve removeUnnecessaryJoints

WebGPURenderer creates shader variants for each skeleton uuid
so we want to unify skeletons as much as possible
If there is a superset skeleton, reuse it

0b5vr committed:
fix: apply patch-package for `NodeMaterial.setupOutput`

0b5vr committed:
Revert "improve: Improve removeUnnecessaryJoints"

This reverts commit 33fc057a43239bb867f8210e315cb375a1b4aa5e.

boneInverse might be different between skeletons
and the new method does not work well with such cases.

0b5vr committed:
Merge branch 'bump-three-r162' into webgpu

0b5vr committed:
fix: update type definition

The `Type of property 'cache' circularly references itself in mapped type` still persists
The error disappears once I bump TypeScript to v5.1.3 or above

0b5vr committed:
fix: tidy up NodeMaterial stuff

revert several examples back to WebGL

`three/examples/jsm/*` now referred by `three/addons/*`; see `tsconfig.json` and `rollup.config.js`

`RotateUV` doesn't work properly with UV Animation Mask Texture on r162. Spin using a handmade rotation matrix instead

0b5vr committed:
fix: Fix rotation order of MToonAnimatedUVNode

tested using VRMC_materials_mtoon_UV_Animation_Test.vrm

0b5vr committed:
feature: add `sameBoneCounts` option to `removeUnnecessaryJoints`

This is an option to decrease initial rendering time when the model has many materials

0b5vr committed:
feature!: MToonNodeMaterialLoaderPlugin

If you want to use three-vrm on WebGPU, use `MToonNodeMaterialLoaderPlugin` by specifying the loader in the option of `VRMLoaderPlugin`.
See the `webgpu-dnd.html` example for details

BREAKING: MToonMaterialLoaderPlugin might put materials other than MToonMaterial (namely, MToonNodeMaterial in this case) to `gltf.userData.vrmMToonMaterials`.
This shouldn't affect many users since the interface of MToonNodeMaterial is almost identical to MToonMaterial

0b5vr committed:
Merge branch 'jsdelivr' into webgpu

0b5vr committed:
chore: change cdn from unpkg to jsDelivr for WebGPU examples

sequel of 4de75c10a6bb433446c740d41d6236620b4f8ea6

0b5vr commented:
Examples are working properly with [r163](https://github.com/mrdoob/three.js/releases/tag/r163).

0b5vr committed:
fix: Patch the ShaderNodeObject circular reference issue on `@types/three`

See: https://github.com/three-types/three-ts-types/pull/744#issuecomment-2036440655

0b5vr committed:
Merge branch 'dev' into webgpu

0b5vr committed:
Merge branch 'dev' into webgpu

0b5vr committed:
Merge branch 'dev' into webgpu

0b5vr committed:
fix: Workaround the `addAssign` issue by using `add` and `assign` separately

We need to fix the issue from the `@types/three` side someday

This change and the previous r164 bump removes all patches to `@types/three`, the commit also removes `patch-package` 🎉

0b5vr committed:
Merge branch 'dev' into webgpu

0b5vr commented:
Ready to merge!

We are going to bump the major since the breaking change occurs.

0b5vr committed:
Merge branch 'dev' into webgpu

0b5vr committed:
chore: reintroduce webgpu-dnd.html

0b5vr committed:
fix: Fix uv animation transform order

The spec says that the animation should not accelerate over time when we apply uv scroll and rotation over time

See: https://github.com/vrm-c/vrm-specification/tree/master/specification/VRMC_materials_mtoon-1.0#transform-order

The GLSL code is also wrong, we have to fix this later

0b5vr committed:
examples: Bump Three.js version of new webgpu examples

0b5vr commented:
日本語で恐縮ですが、NodeMaterialを継承したマテリアルのコードを読む上で、NodeMaterialの動作の流れ（r164現在のもの）を以下のページにまとめてあります。レビューの際にご活用ください。

https://scrapbox.io/0b5vr/Three.js:_NodeMaterial

ke456-png commented on packages/three-vrm-materials-mtoon/src/nodes/MToonNodeMaterialLoaderPlugin.ts:
```diff
@@ -0,0 +1,20 @@
+  public getMaterialType(materialIndex: number): typeof THREE.Material | null {
+    const v1Extension = this._getMToonExtension(materialIndex);
+    if (v1Extension) {
+      return MToonNodeMaterial;
```
[memo]: return MToonNodeMaterial

ke456-png commented on packages/three-vrm-materials-mtoon/src/MToonMaterialLoaderPlugin.ts:
```diff
@@ -146,7 +146,7 @@ export class MToonMaterialLoaderPlugin implements GLTFLoaderPlugin {
   }
 
-  private _getMToonExtension(materialIndex: number): V1MToonSchema.VRMCMaterialsMToon | undefined {
+  protected _getMToonExtension(materialIndex: number): V1MToonSchema.VRMCMaterialsMToon | undefined {
```
[memo] It is necessary to be referenced from MToonNodeMaterialLoaderPlugin.getMaterialType()

ke456-png commented:
The rendering result differs when there is no texture and only a normal map.

examples/feature-test
![image](https://github.com/pixiv/three-vrm/assets/108649297/9e157c92-33a1-41b0-957d-caeda66657b6)

webgpu-feature-test
![image](https://github.com/pixiv/three-vrm/assets/108649297/283b90d6-28d2-47eb-bdcb-25318b8cbfc6)

ke456-png commented:
Perhaps MToonMaterial was wrong?


https://github.com/pixiv/three-vrm/blob/8a8944949e6475e1a418a260dcf443a6f6764289/packages/three-vrm-materials-mtoon/src/MToonMaterial.ts#L607-L614

```
 const useUvInFrag = 
...
      this.uvAnimationMaskTexture !== null ||
      this.normalMap !== null;

```

0b5vr commented:
> The rendering result differs when there is no texture and only a normal map.

It seems to be, yes! The blue ball in the example should have bumps on the surface since the normal map is assigned.
I will fix this in another PR.

ke456-png commented:
If possible, please incorporate this change.
https://github.com/pixiv/three-vrm/commit/25cc2ba3f40c0f80d7da4a51a8e4cd84dc0a79ab

ke456-png commented on packages/three-vrm-materials-mtoon/src/nodes/MToonNodeMaterial.ts:
```diff
@@ -0,0 +1,485 @@
+    // the ordinary diffuseColor setup
+    super.setupDiffuseColor(builder);
+
+    // Set alpha to 1 if it is not transparent
+    if (this.transparent === false) {
+      Nodes.diffuseColor.a.assign(1.0);
+    }
```
I need to confirm the conditions for making it opaque.

https://github.com/pixiv/three-vrm/blob/9e16c3855a5b441cca0f11480c4003de25d2943c/packages/three-vrm-materials-mtoon/src/shaders/mtoon.frag#L871-L873

https://github.com/mrdoob/three.js/blob/687a4caf8c5d633df691a90369c9c185a4acb0fc/src/renderers/webgl/WebGLPrograms.js#L250
`opaque: material.transparent === false && material.blending === NormalBlending && material.alphaToCoverage === false,`

0b5vr replied:
changed that to follow the WebGLRenderer way for now.

https://github.com/pixiv/three-vrm/pull/1384/commits/afe9143d2054097c1b99aae876a77d4bbe58c38c

ke456-png replied:
NICE!

0b5vr committed:
Merge branch 'dev-v3' into webgpu

0b5vr commented:
Resolved conflict with `dev-v3`, it's now ready to build using esbuild.
I don't have to explicitly specify `three/addons/nodes/Nodes.js`, esbuild seems to handle this by only specifying `three`.

0b5vr commented:
> If possible, please incorporate this change. [25cc2ba](https://github.com/pixiv/three-vrm/commit/25cc2ba3f40c0f80d7da4a51a8e4cd84dc0a79ab)

yes I have to indeed

0b5vr committed:
fix: Fix MToon default parameters

equivalent of 25cc2ba3f40c0f80d7da4a51a8e4cd84dc0a79ab

See: https://github.com/pixiv/three-vrm/pull/1384#issuecomment-2124086272

0b5vr commented:
> > If possible, please incorporate this change. [25cc2ba](https://github.com/pixiv/three-vrm/commit/25cc2ba3f40c0f80d7da4a51a8e4cd84dc0a79ab)
> 
> yes I have to indeed

done @ https://github.com/pixiv/three-vrm/pull/1384/commits/36f064fb7653920ec4ff3bf453ac478ba08432dc

0b5vr committed:
feat: emit a compatibility warning if the developer attempts to use MToonNodeMaterial on Three.js r160 or before

0b5vr committed:
feat: MToonNodeMaterial, Move and modify the "set alpha to 1 if it is opaque" part

It's more appropriate to put in `setupOutput` rather than `setupDiffuseColor`

Use the same condition as the `OPAQUE` flag in WebGLRenderer shaders

See: https://github.com/pixiv/three-vrm/pull/1384#discussion_r1609461685

0b5vr commented:
`MToonNodeMaterial` is gonna be compatible only with r161 or higher.
I added a warning if the user attempts to use this in r160 or before.

https://github.com/pixiv/three-vrm/pull/1384/commits/afe9143d2054097c1b99aae876a77d4bbe58c38c

0b5vr committed:
Merge branch 'dev-v3' into webgpu

ke456-png reviewed (approved):
Looks good.
